### ISSUE TITLE
[ENH] Improve categorical variable handling in D1 and D2 of PTF v2.

### ISSUE THREAD

**Issue by PranavBhatP** [jdb78/pytorch-forecasting#1871]
**Is your feature request related to a problem? Please describe.**
The current v2 data pipeline for PTF does not handle 

- Cardinalities: The categorical variables are identified but their cardinalities aren't computed.
- Masking for missing categorical values.
- No embedding support (more specific to model-level implementations like `TimeXer`).

NOTE: This issue is more of a request for enhancement, and does not highlight a failure in the D1-D2 design.

**What is expected?**
From the discussion with @agobbifbk, these were some points that I could infer:
- D1 layer should compute and store categorical cardinalities and pass it as metadata. This is something that can also be computed at D2 level. Design not confirmed yet.
- D2 passes it to model.
- Proper embedding layer initialization for handling categorical data within `TimeXer` and any other such models which might require categorical data inputs. This will build upon the existing implementation in #1836's `data_embedding.py` file.
- Masking of missing categorical variables.

**Comment by fkiraly** [comment]
Not a completely formed opinion, but I want to add to the discussion some minor facets:

* determining the number of categories from an arbitrary input to the D1 constructor can be a resource bottleneck, because in general one would need to go through the entire dataset at least once, if the information is not already available as metadata!
    * this is something we would want to avoid if it is not needed, especially in large scale applications where data may be pre-validated.
    * similarly, it is something that we may want to get from the constructor inputs where possible, instead of getting it from the data. For instance, `pandas.Categorial` does carry the categories and their total number, going through the entire dataset is not required (assuming this is validated).
    * these two points make it a hard design question whether this should be expected metadata output, or optional metadata output, and if optional, then determined by what exactly.
* having a list of encoders, one per variable, feels like brittle design. Should the operation not apply, optimally, to the entire block of categoricals, as a single "multivariate encoder"? Simply producing the output that we would get using the "per-column" application, this would more naturally make use of distribution or parallelization of tensors.

**Comment by agobbifbk** [comment]
I got your points but I have positive vibes, try to replay in the text below:

> Not a completely formed opinion, but I want to add to the discussion some minor facets:
> 
> * determining the number of categories from an arbitrary input to the D1 constructor can be a resource bottleneck, because in general one would need to go through the entire dataset at least once, if the information is not already available as metadata!
>   
>   * this is something we would want to avoid if it is not needed, especially in large scale applications where data may be pre-validated.

I think we need to read data anyway at least once for knowing how many groups you have in the series (without storing them in memory OFC, you can have an incremental label encoder). You will read all the data at some point (aka during the training phase).  I don't see it as a bottleneck. If you want to avoid it (pre-validated data), can we ask to the user the carnality?

>   * similarly, it is something that we may want to get from the constructor inputs where possible, instead of getting it from the data. For instance, `pandas.Categorial` does carry the categories and their total number, going through the entire dataset is not required (assuming this is validated).

In case of `pandasDataFrame` there are for sure better alternatives, but since we are reading some data that can be seen as `np.array` we can always do `np.unique` and incrementally update the encoder.  

>   * these two points make it a hard design question whether this should be expected metadata output, or optional metadata output, and if optional, then determined by what exactly.
> * having a list of encoders, one per variable, feels like brittle design. Should the operation not apply, optimally, to the entire block of categoricals, as a single "multivariate encoder"? Simply producing the output that we would get using the "per-column" application, this would more naturally make use of distribution or parallelization of tensors.

I think the common approach is to create a `ModuleList`, with one embedding per categorical variable as for example IBM is doing  in TTMs see [here](https://github.com/ibm-granite/granite-tsfm/blob/a944109718aa79c72abdd441be8663e326862646/tsfm_public/models/tinytimemixer/modeling_tinytimemixer.py#L120) for reference. They use all the full tensor, I prefer to sum up to along the number of categorical features for not having too much parameters in the following layers.
Personally, I'm investigating alternatives to this, but I think that the one you can find in TTMs is the most used approach. 

BTW, totally available for investigating a more efficient MultivariateEncoder layer!!!

**Comment by fkiraly** [comment]
> If you want to avoid it (pre-validated data), can we ask to the user the carnality?

Yes, one option - better option imo is if whatever is input to the init already carries that information as metadata itself.

> In case of pandasDataFrame there are for sure better alternatives, but since we are reading some data that can be seen as `np.array` we can always do `np.unique` and incrementally update the encoder.

The `np.unique` is exactly what I would prefer to avoid - on scale, you have to run through all chunks, linear time costs can already be high!

> Personally, I'm investigating alternatives to this, but I think that the one you can find in TTMs is the most used approach.

Ok, if we are not reinventing the wheel I will not protest.

**Comment by fkiraly** [comment]
After our discussion today, I realized a number of important points:

* categorical data in `torch` is represented typically as `int64` / `long` subtype. `torch` native categorical encoders and feature extractors assume this as input and will fail on `float`; storage and handling is also more efficient as `int64`.
* even if we would encode categoricals in the D2 layers, it is not possible in the current design, since everything is cast to `float`.
* Hence, D1 needs at least an option to preserve categoricals.
* It is also a reasonable design choice to include encoding in D1, although not a stringently necessary one.

Thoughts about possible design choices from the above below.

**Comment by agobbifbk** [comment]
One more thing: the encoder should be `fitted` during the initialization part (need to see all the data, not at once, we can think something like incremental Ordinal Encoder) and transform the data on the fly in the `get_item`.

**Comment by fkiraly** [comment]
I do not like that we are fitting the encoder.

I think we should enforce the user to use `pandas.Categorical`. This type knows all the categories it has, you can obtain them from `categories`.

This way:
* no looping through the data needed
* we can use a suitable `torch` encoder

**Comment by agobbifbk** [comment]
We can do it in the case of the basic D1 layer where all the data are loaded in RAM, but it won't work in case of a different D1 (for example multiple csv). Moreover, even you cast the categorical columns as `category` and extract the indexes, what happens if you load a new csv? If you cast again as categorical variables you are not sure that it uses the same mapping as used on the whole timeseries, this is why the fitting of an encoder ensure a proper mapping. For sure you can extract the encoder from the `pandas` encoder (in case of simple D1 layer) but if we think to a more general D1 layer we need to take into account that the encoding part needs to be setup properly, do you agree?

**Comment by fkiraly** [comment]
> Moreover, even you cast the categorical columns as category and extract the indexes, what happens if you load a new csv?

We would require that the user passes the category information at construction - it is not inferred from the csvs.

> but if we think to a more general D1 layer we need to take into account that the encoding part needs to be setup properly, do you agree?

Yes - indeed. Everything we change in the D1 output API indeed needs to be changed in *all* D1 classes, and that is exactly what is worrying me! Namely that we make the extension contract overcomplicated, or even impossible/unclear.

The solution that I see is that:

* a specific D1 is compliant already if it always produces only floats - categorical handling is optional
* for D1 that handle categories, it is encouraged that the categories are already queried at construction, the user has to pass this.

**Comment by agobbifbk** [comment]
I take the vacation for clear my mind about this, I think there are more urgent stuff to decide RN.

I do not remember if there is an open issue about how to deal with categorical variables by model side, I leave here one possible solution based on what we do in DSIPTS:
https://colab.research.google.com/drive/188qVS0nBT9tficyEhEVI_vbi8mykCkns?usp=sharing

**Comment by Saptami191** [comment]
@PranavBhatP  @fkiraly  I’ve spent time reviewing the current issue and related v2 API discussions.I checked https://colab.research.google.com/drive/188qVS0nBT9tficyEhEVI_vbi8mykCkns?usp=sharing. I understand that the pipeline currently doesn’t compute or propagate categorical cardinalities, embeds categories, or handle masking for missing values, and that the D1 → D2 metadata flow needs enhancement to support this.The goal here is to support dense representations for categorical variables (e.g., using torch.nn.Embedding), which requires preserving integer category indices and knowing cardinalities so embedding tables can be initialized correctly. Dense embeddings map categorical values to continuous vectors and are a standard pattern in PyTorch for handling high-cardinality categoricals. My proposed idea is to update D1 to optionally preserve categorical indices (torch.long) and include optional metadata for cardinalities, then pass this through D2 to models so that models supporting categorical embeddings can use them. I’d like to work on this improvement and will start by proposing a design before implementing it.
